fix: support multimodal models with non-standard Jinja2 tags

keivenchang opened the pull request:
#### Overview:

Enables support for multimodal models (like LLaVA) that use non-standard Jinja2 template tags unrecognized by minijinja. Implements a placeholder replacement strategy where the Rust frontend transforms custom tags for validation, and the Python backend restores them before vLLM processing.

#### Details:

- Add `replace_non_standard_blocks()` in `formatters.rs` to convert `{% generation %}` → `__JINJA_BLOCK_GENERATION__` for minijinja validation
- Restore placeholders back to original Jinja2 tags in `chat_processor.py` before passing to vLLM's custom AssistantTracker extension
- Make `num_hidden_layers` optional in model card parsing for multimodal models that omit it in `text_config`
- Enhance `eos_token_id` parsing to handle both single integer values and arrays in `generation_config.json`
- Reduce Cargo.lock dependencies

#### Where should the reviewer start?

Start with `lib/llm/src/preprocessor/prompt/template/formatters.rs` to see the placeholder replacement logic, then review `components/src/dynamo/vllm/multimodal_utils/chat_processor.py` to see how placeholders are restored before vLLM processing.

Before the fix:
```
cd examples/backends/vllm/launch/
./agg_multimodal.sh --model llava-hf/llava-1.5-7b-hf
# and then you see this error
2025-11-14T03:20:18.481757Z ERROR dynamo_llm::discovery::watcher: Error adding model from discovery model_name="llava-hf/llava-1.5-7b-hf" namespace="dynamo" error="build_routed_pipeline: syntax error: unknown statement generation (in default:1)"
```

After the fix, you can test completions endpoint (proper format for LLaVA)
```
curl -X POST http://localhost:8000/v1/completions \
-H 'Content-Type: application/json' \
-d '{
"model": "llava-hf/llava-1.5-7b-hf",
"prompt": "USER: Tell me 3 knock knock jokes. ASSISTANT:",
"max_tokens": 150
}' | jq
```


/coderabbit profile chill

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  - Enhanced support for multimodal language models with flexible configuration parsing.
  - Improved template processing for better compatibility with various model configurations.

* **Bug Fixes**
  - Increased robustness of model configuration extraction to handle edge cases and variant data formats.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

KrishnanPrash commented on components/src/dynamo/vllm/multimodal_utils/chat_processor.py:
```diff
@@ -146,6 +146,18 @@ async def preprocess(self, raw_request: ChatCompletionRequest) -> PreprocessResu
+                r"__JINJA_BLOCK_([A-Z_]+)__",
+                lambda m: "{% " + m.group(1).lower() + " %}",
+                chat_template,
+            )
```
From my understanding vLLM relies on `transformers` library to apply chat templates. Here is a relevant snippet from the transformers library:
```python
@lru_cache
def _compile_jinja_template(chat_template):
    if not is_jinja_available():
        raise ImportError(
            "apply_chat_template requires jinja2 to be installed. Please install it using `pip install jinja2`."
        )

    class AssistantTracker(Extension):
        # This extension is used to track the indices of assistant-generated tokens in the rendered chat
        tags = {"generation"}
        
        def __init__(self, environment: ImmutableSandboxedEnvironment):
            # The class is only initiated by jinja.
            ...

        def parse(self, parser: jinja2.parser.Parser) -> jinja2.nodes.CallBlock:
            lineno = next(parser.stream).lineno
            body = parser.parse_statements(["name:endgeneration"], drop_needle=True)
            return jinja2.nodes.CallBlock(self.call_method("_generation_support"), [], [], body).set_lineno(lineno)

        @jinja2.pass_eval_context
        def _generation_support(self, context: jinja2.nodes.EvalContext, caller: jinja2.runtime.Macro) -> str:
            rv = caller()
            ...
            return rv
        
    jinja_env = ImmutableSandboxedEnvironment(
        trim_blocks=True, lstrip_blocks=True, extensions=[AssistantTracker, jinja2.ext.loopcontrols]
    )
    jinja_env.filters["tojson"] = tojson
    jinja_env.globals["raise_exception"] = raise_exception
    jinja_env.globals["strftime_now"] = strftime_now
    return jinja_env.from_string(chat_template)
```

keivenchang replied:
@KrishnanPrash Nice job correcting CodeRabbit!

KrishnanPrash commented:
Thank you for fixing the EPD flow.  Just a few comments, about the other multimodal flow. 

<details>
<summary>My mental model for multimodal models in dynamo+vLLM: </summary>

- Dyanmo+vLLM EPD Path:
  - This flow is invoked with the command `python -m dynamo.vllm --model <MODEL_NAME> --multiomodal-processor`
  - The model is registered with `register_llm(ModelInput.Text, …)`. This means the inference request is accepted and passed as-is to the backend. 
  - We skip pre-processing (applying chat template + tokenization) in Rust frontend
  - We rely on vLLM to internally handle this [[Ref](https://github.com/ai-dynamo/dynamo/blob/ed8cd590a104a9b5f4ca9009e1a2ff543eae4241/components/src/dynamo/vllm/multimodal_utils/chat_processor.py#L138-L168)].
- Dyanmo+vLLM Normal Path:
  - This flow is invoked with the command `python -m dynamo.vllm --model <MODEL_NAME>`
  - This model is registered with `register_llm(ModelInput.Tokens, …)`. This means that the inference request is accepted, the pre-processing happens on the frontend (chat template + tokenization), and we give the backend a PreprocessedRequest Object.
  - Dynamo Rust frontend applies chat template + tokenize
</details>

If we use the Normal Path  in Dynamo + vLLM  with this current PR, an inference request like this:
```bash
 curl -X POST http://localhost:8000/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model": "llava-hf/llava-1.5-7b-hf",
  "messages": [
    {"role": "user", "content": "Who is Jensen Huang?"},
    {"role": "assistant", "content": "Answer with clear detail. Do not omit details."}
  ]
}' | jq
```
After the application of the chat template, the rendered prompt will look like this:
```bash
USER: Who is Jensen Huang? ASSISTANT: __JINJA_BLOCK_GENERATION__Answer with clear detail. Do not omit details. __JINJA_BLOCK_ENDGENERATION__
```

The same cleanup that we do for the epd flow, in my opinion, is needed for the regular flow as well. But instead of re-introducing the custom block tags, we need to remove them all together. 

This could look something like this:
In `oai.rs`:
```rust
fn render() {
   ....
   let rendered = tmpl.render(&ctx)?;
   
   let cleaned = super::formatters::remove_placeholder_blocks(&rendered);

   Ok(cleaned)
}
```

In `formatters.rs`:
```rust
fn remove_placeholder_blocks(rendered: &str) -> String {
    use regex::Regex;

    let re = Regex::new(r"__JINJA_BLOCK_[A-Z_]+__").unwrap();
    re.replace_all(rendered, "").to_string()
}
```
Or we can make it more specfic to only remove the generation tags? Let me know what you think.

KrishnanPrash commented:
Put up #4380 with some additional changes I needed to get this working. Still need to clean the PR up and do some more testing. Will ping you when it is ready for review.

keivenchang commented on lib/llm/src/preprocessor/prompt/template/formatters.rs:
```diff
@@ -9,6 +9,67 @@ use either::Either;
+
+        if !standard_keywords.contains(&tag_name) {
+            // Non-standard tag (e.g., vLLM's {% generation %}) - replace with placeholder
+            let placeholder = format!("__JINJA_BLOCK_{}", tag_name.to_uppercase());
```
Fixed

keivenchang committed:
fix: Handle two non Jinja2 tags in chat templates

- Add remove_known_non_jinja2_tags to strip {% generation %} tags before minijinja validation
- Fixes LLaVA and other models using vLLM-specific template extensions
- Make num_hidden_layers optional in model card parsing for multimodal compatibility
- Handle eos_token_id as single value or array in generation_config.json

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

GuanLuo reviewed (commented):
Overall LGTM. But as Graham brought it up, I believe the multi-modal path does register the model with `ModelInput.Text` as it needs to do something with the raw text (unless this part has been ported to the frontend preprocessor), I am curious if the chat template validation is performed, in the frontend, regardless of the model input type.

KrishnanPrash reviewed (approved):
Looks good to me 👍 . Great Work Here 💯

rmccorm4 commented on lib/llm/src/preprocessor/prompt/template/formatters.rs:
```diff
@@ -9,6 +9,21 @@ use either::Either;
+        .replace("{% generation %}", "")
+        .replace("{% endgeneration %}", "")
+}
+
```
FWIW it also looks like mistralrs also simply removes the generation tags: https://github.com/EricLBuehler/mistral.rs/blob/2bcf0e9e3a0c688786cfa3142e2b18844a7f3447/mistralrs-core/src/pipeline/chat_template.rs#L318-L322

keivenchang commented on lib/llm/src/preprocessor/prompt/template/oai.rs:
```diff
@@ -326,7 +326,24 @@ impl OAIPromptFormatter for HfTokenizerConfigJsonFormatter {
+        //
+        // Note: Non-standard Jinja2 tags like {% generation %} serve different purposes in different paths:
+        // - Normal Path (ModelInput.Tokens): Rust frontend renders the template, so these tags are
+        //   not needed in the output. We remove the placeholders to produce clean text.
```
Removed this change (no change), not relevant anymore.

keivenchang committed:
Add reference to Mistral.rs also stripping generation tags

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
